tracer-transformers/src/Control/Tracer/Transformers/Synopsizer.lhs

```diff
@@ -0,0 +1,99 @@
+\label{code:Control.Tracer.Transformers.WithThreadAndTime}
+
+%if style == newcode
+\begin{code}
+{-|
+Module: Synopsizer
+
+Synopsize runs of repeated events, as an initial message with a summary at the end.
+-}
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE LambdaCase     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE MultiWayIf     #-}
+{-# LANGUAGE RankNTypes     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Control.Tracer.Transformers.Synopsizer
+    (
+    -- * transformer
+      mkSynopsizer
+    , Synopsized(..)
+    ) where
+
+import           Control.Monad.IO.Class (MonadIO (..))
+import           Data.IORef (newIORef,readIORef, writeIORef)
+
+import           Control.Tracer (Tracer (..), traceWith)
+
+\end{code}
+%endif
+
+\begin{code}
+-- | State required to track and summarise repeats.
+data SynopsizerState a
+  = SynopsizerState
+    { ssRepeats  :: {-# UNPACK #-} !Int
+    , ssLast     :: !(Maybe a)
+    }
+
+-- | Abstract representation of a message stream with synopses.
+data Synopsized a
+  = One !a
+    -- ^ Normal, unsynopsized message.
+  | Many {-# UNPACK #-} !Int !a
+    -- ^ Synopsis for a specified number of messages similar to a previous one.
+    --   Note that several 'Many' messages can follow the initial 'One' message,
+    --   because of configurable synopsis overflow.
+
+-- | Generic Tracer transformer.  Will omit consequent runs of similar log objects,
+--   with similarity defined by a supplied predicate.  A run of similar log
+--   objects is terminated by one of two events:
+--     1. a dissimilar event, or
+--     2. reaching the specified overflow limit for omitted messages.
+--
+--   At the end of a run of similar objects, a summary of omission is emitted.
+--
+--   Caveat 1:  the supplied criterion is expected to define a transitive relation.
+--
+--   Caveat 2:  the resulting tracer has state, which is lost upon trace termination.
+--   This means that if the trace ends with a run of similar messages, this will
+--   not be reflected in the trace itself, as observed by the backends
+--   -- they'll only receive the first message of the run.
+--
+--   WARNING:  the resulting tracer is not thread-safe!
+mkSynopsizer :: forall m a
+              . (MonadIO m)
+             => (a -> a -> Bool)
```

Comment:
if the predicate becomes `((Int,a) -> a -> Bool)` with the pair (count, prev value), we could actually show the messages and reset the counter every _n_ occurencies.

Reply:
@CodiePP, I think the Synopsizer needs to know two separate things here:

1. whether the similarity test was successful (i.e. whether to increase the internal counter), and
2. whether we need to "print something already", which is either a summary or the current message.

So that doesn't fit into a `Bool` anymore, and so we'll need to introduce a richer sum type.

Reply:
if the predicate is true then I would assume that the counter is reset. A situation where you print the current state without resetting the counter is counter-intuitive.

Reply:
Ah, you are completely right, thank you Alexander!

Reply:
Implemented, with a caveat -- see https://github.com/input-output-hk/iohk-monitoring-framework/pull/466#issuecomment-563952178